lib/db/schema/patch-048-049.sql

```diff
@@ -0,0 +1,187 @@
+
+    INSERT INTO accounts(
+        uid,
+        normalizedEmail,
+        email,
+        emailCode,
+        emailVerified,
+        kA,
+        wrapWrapKb,
+        authSalt,
+        verifierVersion,
+        verifyHash,
+        verifierSetAt,
+        createdAt,
+        locale
+    )
+    VALUES(
+        inUid,
+        LOWER(inNormalizedEmail),
+        inEmail,
+        inEmailCode,
+        inEmailVerified,
+        inKA,
+        inWrapWrapKb,
+        inAuthSalt,
+        inVerifierVersion,
+        inVerifyHash,
+        inVerifierSetAt,
+        inCreatedAt,
+        inLocale
+    );
+
+    INSERT INTO emails(
+        normalizedEmail,
+        email,
+        uid,
+        emailCode,
+        isVerified,
+        isPrimary,
+        createdAt
+    )
+    VALUES(
+        LOWER(normalizedEmail),
+        inEmail,
+        inUid,
+        inEmailCode,
+        inEmailVerified,
+        true,
+        inCreatedAt
+    );
+
+    COMMIT;
+END;
+
+CREATE PROCEDURE `accountEmails_2` (
+    IN `inUid` BINARY(16)
+)
+BEGIN
+    DROP TEMPORARY TABLE IF EXISTS tempUserEmails;
+    CREATE TEMPORARY TABLE IF NOT EXISTS tempUserEmails AS
+    (SELECT
+        normalizedEmail,
+        email,
+        uid,
+        emailCode,
+        isVerified,
+        isPrimary,
+        verifiedAt,
+        createdAt
+    FROM
+        emails
+    WHERE
+        uid = inUid);
+
+    SET @hasPrimary = 0;
+    SELECT COUNT(*) INTO @hasPrimary FROM tempUserEmails WHERE uid = inUid AND isPrimary = true;
+    IF @hasPrimary = 0 THEN
+        INSERT INTO tempUserEmails
+        SELECT
+            normalizedEmail,
+            email,
+            uid,
+            emailCode,
+            emailVerified AS isVerified,
+            TRUE AS isPrimary,
+            createdAt AS verifiedAt,
+            createdAt
+        FROM
+            accounts a
+        WHERE
+            uid = inUid;
+    END IF;
+
+    SELECT * FROM tempUserEmails ORDER BY isPrimary=true DESC;
+END;
+
+CREATE PROCEDURE `forgotPasswordVerified_6` (
+    IN `inPasswordForgotTokenId` BINARY(32),
+    IN `inAccountResetTokenId` BINARY(32),
+    IN `inTokenData` BINARY(32),
+    IN `inUid` BINARY(16),
+    IN `inCreatedAt` BIGINT UNSIGNED
+)
+BEGIN
+    DECLARE EXIT HANDLER FOR SQLEXCEPTION
+    BEGIN
+        -- ERROR
+        ROLLBACK;
+        RESIGNAL;
+    END;
+
+    START TRANSACTION;
+
+    -- Since we only ever want one accountResetToken per uid, then we
+    -- do a replace - generally due to a collision on the unique uid field.
+    REPLACE INTO accountResetTokens(
+        tokenId,
+        tokenData,
+        uid,
+        createdAt
+    )
+    VALUES(
+        inAccountResetTokenId,
+        inTokenData,
+        inUid,
+        inCreatedAt
+    );
+
+    DELETE FROM passwordForgotTokens WHERE tokenId = inPasswordForgotTokenId;
+
+    UPDATE accounts SET emailVerified = true WHERE uid = inUid;
+    UPDATE emails SET isVerified = true WHERE uid = inUid AND isPrimary = true;
```

Comment:
We'll have to think carefully about this...we don't know which email address was actually used to receive and confirm the reset link, so we need to be very sure that there's no way they could have received it at a secondary email and then have us mark the primary as verified.

I *think* that's OK because you cannot have an unverified primary and a verified secondary, but the logic for enforcing that is far away from the code here that depends on it.  I wonder if it's worth doing an additional check here, and only marking the email verified if the account as a single address and it's an unverified primary.